vars/githubPr.groovy

```diff
@@ -0,0 +1,173 @@
+def isPr() {
+  return !!(env.ghprbPullId && env.ghprbPullLink && env.ghprbPullLink =~ /\/elastic\/kibana\//)
+}
+
+def getLatestBuildComment() {
+  return getComments()
+    .reverse()
+    .find { it.user.login == 'elasticmachine' && it.body =~ /<!--PIPELINE/ }
+}
+
+def getBuildInfoFromComment(commentText) {
+  def matches = commentText =~ /(?ms)<!--PIPELINE(.*?)PIPELINE-->/
+  if (!matches || !matches[0]) {
+    return null
+  }
+
+  return toJSON(matches[0][1].trim())
+}
+
+def getLatestBuildInfo() {
+  return getLatestBuildInfo(getLatestBuildComment())
+}
+
+def getLatestBuildInfo(comment) {
+  return comment ? getBuildInfoFromComment(comment) : null
+}
+
+def createBuildInfo() {
+  return [
+    status: buildUtils.getBuildStatus(),
+    url: env.BUILD_URL,
+    number: env.BUILD_NUMBER,
+    commit: getCommitHash()
+  ]
+}
+
+def getHistoryText(builds) {
+  if (!builds || builds.size() < 1) {
+    return ""
+  }
+
+  def list = builds
+    .reverse()
+    .collect { build ->
+      if (build.status == "SUCCESS") {
+        return "* :green_heart: [Build #${build.number}](${build.url}) succeeded ${build.commit}"
+      } else {
+        return "* :broken_heart: [Build #${build.number}](${build.url}) failed ${build.commit}"
+      }
+    }
+    .join("\n")
+
+  return "### History\n${list}"
+}
+
+def getNextCommentMessage(previousCommentInfo = [:]) {
+  info = previousCommentInfo ?: [:]
+  info.builds = previousCommentInfo.builds ?: []
+
+  def messages = []
+
+  if (buildUtils.getBuildStatus() == 'SUCCESS') {
+    messages << """
+      ## :green_heart: Build Succeeded
+      * [continuous-integration/kibana-ci/pull-request](${env.BUILD_URL})
+      * Commit: ${getCommitHash()}
+    """
+  } else {
+    messages << """
+      ## :broken_heart: Build Failed
+      * [continuous-integration/kibana-ci/pull-request](${env.BUILD_URL})
+      * Commit: ${getCommitHash()}
+    """
+  }
+
+  if (info.builds && info.builds.size() > 0) {
+    messages << getHistoryText(info.builds)
+  }
+
+  messages << "To update your PR or re-run it, just comment with:\n`@elasticmachine merge upstream`"
+
+  info.builds << createBuildInfo()
+
+  messages << """
+    <!--PIPELINE
+    ${toJSON(info).toString()}
+    PIPELINE-->
+  """
+
+  return messages
+    .findAll { !!it } // No blank strings
+    .collect { it.stripIndent().trim() }
+    .join("\n\n")
+}
+
+def withGithubCredentials(closure) {
+  withCredentials([
+    string(credentialsId: '2a9602aa-ab9f-4e52-baf3-b71ca88469c7', variable: 'GITHUB_TOKEN'),
+  ]) {
+    closure()
+  }
+}
+
+def postComment(message) {
+  if (!isPr()) {
+    error "Trying to post a GitHub PR comment on a non-PR or non-elastic PR build"
+  }
+
+  withGithubCredentials {
+    return githubApi.post("repos/elastic/kibana/issues/${env.ghprbPullId}/comments", [ body: message ])
+  }
+}
+
+def getComments() {
+  withGithubCredentials {
+    return githubApi.get("repos/elastic/kibana/issues/${env.ghprbPullId}/comments")
```

Comment:
Probably want to do `?sort=created&direction=desc` here?

Reply:
Will add that, that seemed to be the default and I didn't even think to check/make it explicit

Reply:
Doesn't look like this method is sortable https://developer.github.com/v3/issues/comments/#list-comments-on-an-issue

Reply:
I created another method in kibana-pipeline-library that uses a Java-based GitHub client to pull comments. The client takes care of paging/etc